use set_buffer_size_near to calculate fixed alsa buffer size

attackgoat opened the pull request:
Fixes #743

Hardware parameters cannot be set to some arbitrary values within the supported range.

Two alternate fixes are:
- Expose the "near" ALSA function to the user
- Return the actual buffer size used

I am not familiar enough with CPAL usage to know if the alternate fixes are needed, but for my use cases they are not.

roderickvd commented:
Agreed, this is the right thing to do. This highly related commit in `librespot` actually combines the two approaches you mention: https://github.com/librespot-org/librespot/commit/d5efb8a620554ced4fda731c72046c4b1af53111

roderickvd commented:
@attackgoat as I became a maintainer today, and agree with your approach, I would like to work with you to getting to a mergable state. Could you update it to:
- take the device default when the "near" function fails? (look at the librespot example for inspiration)
- add a changelog entry

Thanks!

roderickvd commented:
I see that #917 also does this (and more).

attackgoat commented:
> I see that #917 also does this (and more).

That PR seems to be the better direction, although it causes me the same issue with large buffers. For now I've addressed the feedback but we may want to close this PR if #917 is cleaned up.

roderickvd commented:
Thanks a lot @attackgoat. The author of #917 expressed that he is no longer working on it. Would you be willing to merge the two approaches?

attackgoat commented on src/host/alsa/mod.rs:
```diff
@@ -1104,19 +1104,27 @@ fn set_hw_params_from_format(
+    if let BufferSize::Default = buffer_size {
+        // These values together represent a moderate latency and wakeup interval.
+        // Without them, we are at the mercy of the device
+        hw_params.set_period_time_near(25_000, alsa::ValueOr::Nearest)?;
```
The latest commit handles various failures better and relies on the actual ALSA-provided period _size_ after setting _time_.

roderickvd reviewed (changes requested):
Thanks a lot, this is really going in the right direction. Some points to make it even more robust - I've got prior experience where I know that ALSA drivers can be really fickle.

At any time just let me know if and how I can help.

roderickvd reviewed (changes requested):
Well implemented with one final bit of polish! 👏

attackgoat commented on src/host/alsa/mod.rs:
```diff
@@ -1104,22 +1135,95 @@ fn set_hw_params_from_format(
+        };
+
+        // Actual period size
+        let Ok(period_size) =
```
Sounds good - the code has been adjusted to return `BackendSpecificError` in case it cannot set the requested buffer size (default or fixed).

In the future I think it would be helpful to return a better error type that programs could inspect and use to make decisions, because the string-based error is only helpful for the programmer. The `log` crate would be handy here.
